extensions/amp-image-slider/1.0/component.js

```diff
@@ -0,0 +1,607 @@
+
+/**
+ * Using Gestures class throws console error for failed to call initLogConstructor
+ * so we need to explicity call it.
+ */
+initLogConstructor();
+installTimerService(global);
+
+/**
+ * Displays given component with supplied props.
+ * @param {*} props
+ * @param {{current: ?Element}} ref
+ * @return {PreactDef.Renderable}
+ */
+function DisplayAsWithRef({as: Comp = 'div', containerClass, ...rest}, ref) {
+  return (
+    /** Need containerClass to add jss Styling to the component container */
+    <div ref={ref} class={containerClass}>
+      <Comp {...rest} />
+    </div>
+  );
+}
+
+const DisplayAs = forwardRef(DisplayAsWithRef);
+
+/**
+ * @param {!BentoImageSliderDef.Props} props
+ * @param {{current: ?BentoImageSliderDef.Api}} ref
+ * @return {PreactDef.Renderable}
+ */
+export function BentoImageSliderWithRef(
+  {
+    displayHintOnce,
+    firstImageAs,
+    firstLabelAs,
+    initialPosition,
+    leftHintAs,
+    rightHintAs,
+    secondImageAs,
+    secondLabelAs,
+    stepSize = 0.1,
+    ...rest
+  },
+  ref
+) {
+  /** Common variables */
+  const gesturesRef = useRef(null);
+  const isEventRegistered = false;
+  const styles = useStyles();
+  /** Common variables */
+  const doc = useRef(null);
+  /** Container Reference */
+  const containerRef = useRef(null);
+
+  /** Image References */
+  const leftImageRef = useRef(null);
+  const rightImageRef = useRef(null);
+
+  /** Label References */
+  const leftLabelWrapperRef = useRef(null);
+  const leftLabelRef = useRef(null);
+  const rightLabelWrapperRef = useRef(null);
+  const rightLabelRef = useRef(null);
+
+  /** Mask References */
+  const leftMaskRef = useRef(null);
+  const rightMaskRef = useRef(null);
+
+  /** Bar Reference */
+  const barRef = useRef(null);
+
+  /** Arrow References */
+  const leftHintArrowRef = useRef(null);
+  const rightHintArrowRef = useRef(null);
+
+  /** Hint Wrapper References */
+  const leftHintWrapper = useRef(null);
+  const rightHintWrapper = useRef(null);
+
+  /** Hint Body References */
+  const leftHintBodyRef = useRef(null);
+  const rightHintBodyRef = useRef(null);
+
+  /** Unlisten Handlers for Mouse */
+  const unlistenMouseUp = useRef(null);
+  const unlistenMouseDown = useRef(null);
+  const unlistenMouseMove = useRef(null);
+
+  /** Unlisten Handlers for Keyboard */
+  const unlistenKeyDown = useRef(null);
+
+  /** Animate Hint Flag */
+  const [hideHint, setHideHint] = useState(false);
+
+  const animateHideHint = useCallback(() => {
+    setHideHint(true);
+  }, [setHideHint]);
+
+  const animateShowHint = useCallback(() => {
+    setHideHint(false);
+  }, [setHideHint]);
+
+  /** Listen / Unlisten Handler */
+  /**
+   * Unlisten a listener and clear. If null, does nothing
+   * @param {?UnlistenDef} unlistenHandle
+   * @private
+   */
+  const unlisten = useCallback((unlistenHandle) => {
+    if (unlistenHandle) {
+      unlistenHandle();
+      unlistenHandle = null;
```

Comment:
Is this suppose to "delete" the `unlistenHandle`?

The original argument will not be mutated by this. We'll either want to pass in the ref or set the ref with whatever this function returns?

Reply:
Thanks, I have updated the logic by passing the reference to the `unlisten` function.